src/multisampler.jl

```diff
@@ -0,0 +1,141 @@
+struct Sample
+    g::DiGraph
+    τ::Float64
+    dir::Int8
+    total::Int32 
+    scoreval::Float64
+end
+
+struct Action
+    τ::Float64 
+    func::Function
+    args::Tuple{Vararg{Any}}
+end
+
+function expcoldness(τ, k=0.001)
+    return exp(k*τ)
+end
+
+function Dexpcoldness(τ, k=0.001)
+    return k*exp(k*τ)
+end
+
+function init(_, _, nextτ, g, dir, total, scoreval)
+    return Sample(g, nextτ, dir, total, scoreval) 
+end
+
+function applyup(samplers, i, nextτ, x, y, T, Δscoreval)
+    prevsample = last(samplers[i])
+    g = next_CPDAG(prevsample.g, :up, x, y, T)
+    return Sample(g, nextτ, prevsample.dir, prevsample.total+1, prevsample.scoreval + Δscoreval)
+end
+
+function applydown(samplers, i, nextτ, x, y, H, Δscoreval)
+    prevsample = last(samplers[i])
+    g = next_CPDAG(prevsample.g, :down, x, y, H)
+    return Sample(g, nextτ, prevsample.dir, prevsample.total-1, prevsample.scoreval + Δscoreval) 
+end
+
+function applyflip(samplers, i, nextτ)
+    prevsample = last(samplers[i])
+    return Sample(prevsample.g, nextτ, -1*prevsample.dir, prevsample.total, prevsample.scoreval)
+end
+
+function applycopy(samplers, i, nextτ, j)
+    sample = last(samplers[i])
+    return Sample(sample.g, nextτ, sample.dir, sample.total, sample.scoreval)
+end
+
+# for starters without turn move
+function sampleaction(samplers, i, M, balance, prior, score, σ, ρ, κ) 
+    # preprocess 
+    prevsample = last(samplers[i])
+    sup, sdown, Δscorevalup, Δscorevaldown, argsup, argsdown = count_moves_new(prevsample.g, κ, balance, prior, score, expcoldness(prevsample.τ), prevsample.total)
+
+    # propose moves
+    λdir = prevsample.dir == 1 ? sup : sdown 
+    λupdown = sup + sdown 
+    λflip = max(prevsample.dir*(-sup + sdown), 0.0)
+    # TODO: does this make any sense?
+    λterm = Dexpcoldness(prevsample.τ) * expcoldness(prevsample.τ) * prevsample.scoreval # TODO: prior
+
+    Δτdir = randexp()/(ρ*λdir)
+    Δτupdown = randexp()/(σ*λupdown)
+    Δτflip = randexp()/(ρ*λflip)
+    Δτterm = randexp()/(λterm) 
+
+    Δτmin = min(Δτdir, Δτupdown, Δτflip, Δτterm)
+
+    if Δτdir == Δτmin 
+        if prevsample.dir == 1
+            return Action(prevsample.τ + Δτdir, applyup, (argsup..., Δscorevalup))
+        else 
+            return Action(prevsample.τ + Δτdir, applydown, (argsdown..., Δscorevaldown))
+        end
+    end
+
+    if Δτupdown == Δτmin
+        λup = sup
+        if rand() < λup/λupdown
+            return Action(prevsample.τ + Δτupdown, applyup, (argsup..., Δscorevalup))
+        else 
+            return Action(prevsample.τ + Δτupdown, applydown, (argsdown..., Δscorevaldown))
+        end
+    end
+    
+    if Δτflip == Δτmin 
+        return Action(prevsample.τ + Δτflip, applyflip, ())
+    end
+
+    if Δτterm == Δτmin 
+        return Action(prevsample.τ + Δτterm, applycopy, (rand(1:M),)) 
```

Comment:
the sample [i] is actually part of the representative sample from the annealed distribution and is therefore included in the possible choices.